New Resource: `azurerm_arc_kubernetes_cluster`

ms-zhenhua opened the pull request:
Fixes #12612
swagger: https://github.com/Azure/azure-rest-api-specs/blob/main/specification/hybridkubernetes/resource-manager/Microsoft.Kubernetes/stable/2021-10-01/connectedClusters.json

Test Case Result:
![image](https://user-images.githubusercontent.com/76987228/227434913-6c63bf61-9700-4888-8a5f-dd6c19f5e3f1.png)

tombuildsstuff commented on internal/services/hybridkubernetes/hybrid_kubernetes_connected_cluster_resource_test.go:
```diff
@@ -0,0 +1,193 @@
+		}
+		return nil, fmt.Errorf("retrieving %s: %+v", id, err)
+	}
+	return utils.Bool(true), nil
```
```suggestion
	return utils.Bool(resp.Model != nil), nil
```

tombuildsstuff commented on internal/services/hybridkubernetes/hybrid_kubernetes_connected_cluster_resource_test.go:
```diff
@@ -0,0 +1,193 @@
+
+resource "azurerm_hybrid_kubernetes_connected_cluster" "import" {
+  name                         = azurerm_hybrid_kubernetes_connected_cluster.test.name
+  resource_group_name          = azurerm_resource_group.test.name
```
```suggestion
  resource_group_name          = azurerm_hybrid_kubernetes_connected_cluster.test.resource_group_name
```

tombuildsstuff commented on internal/services/hybridkubernetes/hybrid_kubernetes_connected_cluster_resource_test.go:
```diff
@@ -0,0 +1,193 @@
+  name                         = "acctest-k-%d"
+  resource_group_name          = azurerm_resource_group.test.name
+  agent_public_key_certificate = "MIICCgKCAgEAsSpALlON3394ysLQdRSy6cCBwL08NgZp7c1xsy0kQH/wHuixfoCwtL1OZ0a5kqod9vE6L8ICsXAE+iEdU1OspcJxL9J/gSyiOCMYPUabbYRXFy5x258RRLtn60NoaqcaDW+Z80HLwJOMECdJ/yDkuuNbnL0J2cyR8/WXjoeee8cG52QmDuxB6a4ROOushroIE2NS3FuhJh3b3Ddj+NU3gfbgIjjCMJQDpJWqUi/68lB33K1FQwuY1CCT8rKC5dDRKnlTgEWtnhBBly1D+C9GPTh7l9NTxcAEcBKo1ZIkMcixcS+gTcsTtDRRwuTtt8kybohfRMvmBA/S9bkD6cCaGJMe8YerOyJsD ...
+  distribution                 = "kind"
+
+  identity {
+    type = "SystemAssigned"
+  }
+  infrastructure = "generic"
```
are either of these are valid values?

ms-zhenhua replied:
Yes, I've test it against a Kubernetes deployed on Kind following this guidance(https://docs.microsoft.com/en-us/azure/azure-arc/kubernetes/quickstart-connect-cluster?tabs=azure-cli).

tombuildsstuff commented on internal/services/hybridkubernetes/registration.go:
```diff
@@ -0,0 +1,29 @@
+// WebsiteCategories returns a list of categories which can be used for the sidebar
+func (r Registration) WebsiteCategories() []string {
+	return []string{
+		"HybridKubernetes",
```
per the product name, is this Azure Arc? https://docs.microsoft.com/en-us/azure/azure-arc/kubernetes/

```suggestion
		"Arc",
```

ms-zhenhua replied:
Will rename it to "ArcKubernetes"

tombuildsstuff commented on internal/services/hybridkubernetes/registration.go:
```diff
@@ -0,0 +1,29 @@
+// SupportedResources returns the supported Resources supported by this Service
+func (r Registration) SupportedResources() map[string]*pluginsdk.Resource {
+	return map[string]*pluginsdk.Resource{
+		"azurerm_hybrid_kubernetes_connected_cluster": resourceHybridKubernetesConnectedCluster(),
```
per the product name, should this be:

```suggestion
		"azurerm_arc_kubernetes_cluster": resourceArcKubernetesCluster(),
```

tombuildsstuff commented on internal/services/hybridkubernetes/resourceids.go:
```diff
@@ -0,0 +1,3 @@
+package hybridkubernetes
+
+//go:generate go run ../../tools/generator-resource-id/main.go -path=./ -name=HybridKubernetesConnectedCluster -id=/subscriptions/12345678-1234-9876-4563-123456789012/resourceGroups/resourceGroup1/providers/Microsoft.Kubernetes/connectedClusters/cluster1
```
this file is not necessary?

```suggestion
```

tombuildsstuff commented on website/docs/r/hybrid_kubernetes_connected_cluster.html.markdown:
```diff
@@ -0,0 +1,98 @@
+  location = "West Europe"
+}
+
+resource "azurerm_hybrid_kubernetes_connected_cluster" "test" {
+  name                         = "example-connected-cluster"
+  resource_group_name          = azurerm_resource_group.test.name
+  agent_public_key_certificate = "xxxxxxxxxxxxxxxxxxx"
+  distribution                 = "gke"
+
+  identity {
+    type = "SystemAssigned"
+  }
+  infrastructure = "gcp"
+  location       = "West Europe"
+
+  tags = {
+    ENV = "Test"
+  }
+}
```
we should reorder this to be:

```suggestion
resource "azurerm_hybrid_kubernetes_connected_cluster" "test" {
  name                         = "example-connected-cluster"
  resource_group_name          = azurerm_resource_group.test.name
  location       = azurerm_resource_group.test.location
  distribution                 = "gke"
  infrastructure = "gcp"
  agent_public_key_certificate = "xxxxxxxxxxxxxxxxxxx"
  
  identity {
    type = "SystemAssigned"
  }

  tags = {
    ENV = "Test"
  }
}
```

tombuildsstuff commented on website/docs/r/hybrid_kubernetes_connected_cluster.html.markdown:
```diff
@@ -0,0 +1,98 @@
+
+The following arguments are supported:
+
+* `name` - (Required) The name which should be used for this HybridKubernetes Connected Cluster. Changing this forces a new HybridKubernetes Connected Cluster to be created.
```
```suggestion
* `name` - (Required) Specifies the name which should be used for this Arc Kubernetes Cluster. Changing this forces a new Arc Kubernetes Cluster to be created.
```

tombuildsstuff commented on website/docs/r/hybrid_kubernetes_connected_cluster.html.markdown:
```diff
@@ -0,0 +1,98 @@
+
+* `name` - (Required) The name which should be used for this HybridKubernetes Connected Cluster. Changing this forces a new HybridKubernetes Connected Cluster to be created.
+
+* `resource_group_name` - (Required) The name of the Resource Group where the HybridKubernetes connected cluster should exist. Changing this forces a new HybridKubernetes connected cluster to be created.
```
```suggestion
* `resource_group_name` - (Required) Specifies the name of the Resource Group where the Arc Kubernetes Cluster should exist. Changing this forces a new Arc Kubernetes Cluster to be created.
```

tombuildsstuff commented on website/docs/r/hybrid_kubernetes_connected_cluster.html.markdown:
```diff
@@ -0,0 +1,98 @@
+
+* `resource_group_name` - (Required) The name of the Resource Group where the HybridKubernetes connected cluster should exist. Changing this forces a new HybridKubernetes connected cluster to be created.
+
+* `agent_public_key_certificate` - (Required) Base64 encoded public certificate used by the agent to do the initial handshake to the backend services in Azure. Changing this forces a new HybridKubernetes connected cluster to be created.
```
```suggestion
* `agent_public_key_certificate` - (Required) Specifies the base64-encoded public certificate used by the agent to do the initial handshake to the backend services in Azure. Changing this forces a new Arc Kubernetes Cluster to be created.
```

tombuildsstuff commented on website/docs/r/hybrid_kubernetes_connected_cluster.html.markdown:
```diff
@@ -0,0 +1,98 @@
+
+* `agent_public_key_certificate` - (Required) Base64 encoded public certificate used by the agent to do the initial handshake to the backend services in Azure. Changing this forces a new HybridKubernetes connected cluster to be created.
+
+* `identity` - (Required) An `identity` block as defined below.
```
```suggestion
* `identity` - (Required) An `identity` block as defined below. Changing this forces a new Arc Kubernetes Cluster to be created.
```

tombuildsstuff commented on website/docs/r/hybrid_kubernetes_connected_cluster.html.markdown:
```diff
@@ -0,0 +1,98 @@
+
+* `identity` - (Required) An `identity` block as defined below.
+
+* `location` - (Required) The Azure Region where the HybridKubernetes connected cluster should exist. Changing this forces a new HybridKubernetes connected cluster to be created.
```
```suggestion
* `location` - (Required) Specifies the Azure Region where the Arc Kubernetes Cluster should exist. Changing this forces a new Arc Kubernetes Cluster to be created.
```

tombuildsstuff commented on website/docs/r/hybrid_kubernetes_connected_cluster.html.markdown:
```diff
@@ -0,0 +1,98 @@
+
+* `location` - (Required) The Azure Region where the HybridKubernetes connected cluster should exist. Changing this forces a new HybridKubernetes connected cluster to be created.
+
+* `distribution` - (Optional) The HybridKubernetes distribution running on this connected cluster. Changing this forces a new HybridKubernetes connected cluster to be created.
```
three things:

1.

```suggestion
* `distribution` - (Optional) Specifies the distribution running on this Arc Kubernetes Cluster. Changing this forces a new Arc Kubernetes Cluster to be created.
```

2. What does this do?
3. Should this be a user-specifiable field? is this computed-only?

ms-zhenhua replied:
Both **distribution** and **infrastructure** are user-specifiable fields. **infrastructure** refers to the cloud provider where the Kubernetes cluster resides (like Azure, aws, gcp). **distribution** refers to the type of the Kubernetes cluster itself(like aks, gke, eks). However, these values need to be manually configured by users. Currently, both of these fields are simple strings only and will not be validated by the back end.

tombuildsstuff commented on website/docs/r/hybrid_kubernetes_connected_cluster.html.markdown:
```diff
@@ -0,0 +1,98 @@
+
+* `distribution` - (Optional) The HybridKubernetes distribution running on this connected cluster. Changing this forces a new HybridKubernetes connected cluster to be created.
+
+* `infrastructure` - (Optional) The infrastructure on which the HybridKubernetes cluster represented by this connected cluster is running on. Changing this forces a new HybridKubernetes connected cluster to be created.
```
same 3 questions here:

```suggestion
* `infrastructure` - (Optional) Specifies the infrastructure on which the Arc Kubernetes Cluster is running on. Changing this forces a new Arc Kubernetes Cluster to be created.
```

2. What does this do?
3. Should this be user-specifiable, or is this a computed-only field?

ms-zhenhua replied:
Answered above.

tombuildsstuff commented on website/docs/r/hybrid_kubernetes_connected_cluster.html.markdown:
```diff
@@ -0,0 +1,98 @@
+
+* `infrastructure` - (Optional) The infrastructure on which the HybridKubernetes cluster represented by this connected cluster is running on. Changing this forces a new HybridKubernetes connected cluster to be created.
+
+* `tags` - (Optional) A mapping of tags which should be assigned to the HybridKubernetes connected cluster.
```
```suggestion
* `tags` - (Optional) A mapping of tags which should be assigned to the Arc Kubernetes Cluster.
```

tombuildsstuff commented on website/docs/r/hybrid_kubernetes_connected_cluster.html.markdown:
```diff
@@ -0,0 +1,98 @@
+
+An `identity` block exports the following:
+
+* `type` - (Required) Specifies the type of Managed Service Identity. Possible values are `SystemAssigned`.
```
```suggestion
* `type` - (Required) Specifies the type of Managed Service Identity assigned to this Arc Kubernetes Cluster. At this time the only possible value is `SystemAssigned`.
```

tombuildsstuff commented on website/docs/r/hybrid_kubernetes_connected_cluster.html.markdown:
```diff
@@ -0,0 +1,98 @@
+
+---
+
+An `identity` block exports the following:
```
```suggestion
An `identity` block supports the following:
```

tombuildsstuff commented on website/docs/r/hybrid_kubernetes_connected_cluster.html.markdown:
```diff
@@ -0,0 +1,98 @@
+
+In addition to the Arguments listed above - the following Attributes are exported:
+
+* `id` - The ID of the HybridKubernetes connected cluster.
```
```suggestion
* `id` - The ID of the Arc Kubernetes Cluster.
```

tombuildsstuff commented on website/docs/r/hybrid_kubernetes_connected_cluster.html.markdown:
```diff
@@ -0,0 +1,98 @@
+
+* `identity` - An `identity` block as defined below.
+
+* `provisioning_state` - The current deployment state of connected clusters.
```
this field should be removed:

```suggestion
```

and the other computed fields should be added:

tombuildsstuff commented on website/docs/r/hybrid_kubernetes_connected_cluster.html.markdown:
```diff
@@ -0,0 +1,98 @@
+
+The `timeouts` block allows you to specify [timeouts](https://www.terraform.io/docs/configuration/resources.html#timeouts) for certain actions:
+
+* `create` - (Defaults to 30 minutes) Used when creating the HybridKubernetes connected cluster.
+* `read` - (Defaults to 5 minutes) Used when retrieving the HybridKubernetes connected cluster.
+* `update` - (Defaults to 30 minutes) Used when updating the HybridKubernetes connected cluster.
+* `delete` - (Defaults to 30 minutes) Used when deleting the HybridKubernetes connected cluster.
```
```suggestion
* `create` - (Defaults to 30 minutes) Used when creating the Arc Kubernetes Cluster. 
* `read` - (Defaults to 5 minutes) Used when retrieving the Arc Kubernetes Cluster.
* `update` - (Defaults to 30 minutes) Used when updating the Arc Kubernetes Cluster.
* `delete` - (Defaults to 30 minutes) Used when deleting the Arc Kubernetes Cluster.
```

tombuildsstuff commented on website/docs/r/hybrid_kubernetes_connected_cluster.html.markdown:
```diff
@@ -0,0 +1,98 @@
+
+## Import
+
+HybridKubernetes connected cluster can be imported using the `resource id`, e.g.
```
```suggestion
Arc Kubernetes Clusters can be imported using the `resource id`, e.g.
```

tombuildsstuff commented on website/docs/r/hybrid_kubernetes_connected_cluster.html.markdown:
````diff
@@ -0,0 +1,98 @@
+HybridKubernetes connected cluster can be imported using the `resource id`, e.g.
+
+```shell
+terraform import azurerm_hybrid_kubernetes_connected_cluster.example /subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/resourceGroup1/providers/Microsoft.Kubernetes/connectedClusters/cluster1
````
```suggestion
terraform import azurerm_arc_kubernetes_cluster.example /subscriptions/00000000-0000-0000-0000-000000000000/resourceGroups/resourceGroup1/providers/Microsoft.Kubernetes/connectedClusters/cluster1
```

tombuildsstuff commented on internal/services/hybridkubernetes/hybrid_kubernetes_connected_cluster_resource.go:
```diff
@@ -0,0 +1,206 @@
+		Update: resourceHybridKubernetesConnectedClusterUpdate,
+		Delete: resourceHybridKubernetesConnectedClusterDelete,
+
+		Timeouts: &pluginsdk.ResourceTimeout{
+			Create: pluginsdk.DefaultTimeout(30 * time.Minute),
+			Read:   pluginsdk.DefaultTimeout(5 * time.Minute),
+			Update: pluginsdk.DefaultTimeout(30 * time.Minute),
+			Delete: pluginsdk.DefaultTimeout(30 * time.Minute),
+		},
```
are these timeouts valid for this service?

ms-zhenhua replied:
Yes, I have already test this resource that each operation will take no more than 10 seconds in normal situation. So I think these values are valid.

ms-zhenhua commented on internal/services/hybridkubernetes/hybrid_kubernetes_connected_cluster_resource.go:
```diff
@@ -0,0 +1,206 @@
+		Identity: *identity,
+		Location: location,
+		Properties: hybridkubernetes.ConnectedClusterProperties{
+			AgentPublicKeyCertificate: d.Get("agent_public_key_certificate").(string),
```
Since it's base64-encoded value, I will add validation for this field and do not need to normalize in a method.

ms-zhenhua commented:
Hi @tombuildsstuff , thank you for reviewing my code. I have updated the code and re-run the test cases with no error. Kindly help review it again. And I've also created a k8s cluster and successfully connected it to Azure. BTW, the Golang lintint failed because of timeout. But it ran successfully on my own repo. Kindly help re-run it. Thanks.
![image](https://user-images.githubusercontent.com/76987228/157014054-c3d18824-846e-4884-a87c-e0aff0ba4e54.png)

ms-zhenhua commented:
Hi @stephybun, I saw your question was [answered](https://github.com/Azure/azure-cli-extensions/issues/4756#issuecomment-1141722260) in this [issue](https://github.com/Azure/azure-cli-extensions/issues/4756). Could this PR be unblocked now?

ageisen2000 commented:
Looking forward to this integration.

ms-zhenhua commented:
Temporarily close this PR to make some modifications. Will reopen it later.

ms-zhenhua commented:
Hi @stephybun, I updated the PR to let it support installing the arc agent. I extracted the code for installing arc agent from [Azure Cli extension](https://github.com/Azure/azure-cli-extensions/blob/main/src/connectedk8s/azext_connectedk8s/custom.py) and the code can be run in a Linux VM through [remote-exec provisioner](https://developer.hashicorp.com/terraform/language/resources/provisioners/remote-exec). Now all test cases can be run automatically. Could you kindly help take another review? Thanks.

stephybun reviewed (changes requested):
Thanks @ms-zhenhua.

In addition to the comments left in-line I still have concerns about how a user can establish connectivity to the Arc resource provisioned through Terraform. 

Although the logic to install the agent can be taken from the Azure CLI, this isn't documented publicly anywhere for us to reference and is left up to the user to go digging for it. Could we get this process documented on the MSFT learn site as well as add a valid example to the examples folder for this resource?

ms-zhenhua commented:
Hi @stephybun, your concerns are reasonable. In fact, there already exists a document for [agent deployment](https://learn.microsoft.com/en-us/azure/azure-arc/kubernetes/conceptual-agent-overview#deploy-agents-to-your-cluster). I have also created a [requirement issue](https://github.com/Azure/azure-rest-api-specs/issues/23247) to ask service team to make the document easier for users to understand and will keep tracking it. I have also added a note to remind users to install agents when creating this resource and added an example to the example folder.
All comments are also resolved. Kindly take another review.

akfmdl commented:
I really need this feature!!

stephybun commented on examples/arckubernetes/README.md:
```diff
@@ -0,0 +1,24 @@
+6. A [Kind Cluster](https://kind.sigs.k8s.io/) in [Linux Virtual Machine](https://registry.terraform.io/providers/hashicorp/azurerm/latest/docs/resources/linux_virtual_machine) by [remote-exec Provisioner](https://developer.hashicorp.com/terraform/language/resources/provisioners/remote-exec)
+7. [Azure Arc agents](https://learn.microsoft.com/en-us/azure/azure-arc/kubernetes/conceptual-agent-overview) in [Kind Cluster](https://kind.sigs.k8s.io/) by [remote-exec Provisioner](https://developer.hashicorp.com/terraform/language/resources/provisioners/remote-exec)
+
+~> **NOTE:** To Connect an existing Kubernetes cluster to Azure Arc, the following conditions must be met:
```
```suggestion
~> **NOTE:** To connect an existing Kubernetes cluster to Azure Arc, the following conditions must be met:

```

stephybun commented on examples/arckubernetes/testdata/install_agent.py:
```diff
@@ -0,0 +1,249 @@
+                         args.location,
+                         args.tenantId,
+                         privateKey,
+                         # private_key_pem2,
```
Remove this comment
```suggestion

```

stephybun commented on examples/arckubernetes/variables.tf:
```diff
@@ -0,0 +1,78 @@
+7dEAMQ5PySRBXWt1FBdn0d55onwZy29GgqHge33/6sn/QJv5Fvsf2peXPqqJcCcE
+O0XEK+LqOWS9ukAzMyxIsXmC0T80qvTizETt0Vrh93WVslIRfqLj0yZJnDiFlsJH
+H31zFU9SivnxtXTUgiOzi7HSPueHEGgTONmnswzyQWwAdTG0NaKaCqpTaCbL1x9O
+FHD8Zq4ubHLWaJoStfyrFVcm4Hi1KGZ2MX5vOnxJLbG4+86Y/80g1kjJQ/s9mfrF
+Zh4MiQ6FD7iCRmAzTaRbn+4P0ToWrw9KTFDaBupudfw0OVfxAQKCAQEA2aLgcKkg
+urH+dNelUupbgQiVn+ctdNbR2AJuDSTH0jOBq6bsaqOymtkBrNHoFBooNZGI9hK3
+zJ7BQHNI60gQ6QJmI8lfPle40s5U281ZKoEQ7g/5NXHjMTZIHnv3pQ5hsdr2510F
+sQkZY7epQjwWb+CIf2cupZop+7P3/RYy2KEY06w8O9P6zjM+HRXqzdW3xMBV9qEe
+sGtGVhoX7SI7tNe6OGEHkQ5nLC1xlEQ/56IH73z7MMik0Zk162IQ2tkW+rsOQUWV
+E0JoqG4E0YW0cRd3VBjsOr75WPMJkjNCxVI8HP0BrnoLoR9Y5NPp9npkbs0PZoSL
+92Cy+HGNOVvazwKCAQEA2K9HGhes30z4vF6D8V2aKt2cfwZW+BRBiwp5Cy1mGMJE
+DYs0uYKeCFyTsrWBiAUVgFi+0hArSSGCgBg+2hqwx/uycrsrqDm3m+35gQt9KnTh
+fC6vCKN/9jbFadvSnBLJgmh9xLOR3wUtL26LIRwLDkOGiQOLEsz0akmj2BYjFugl
+ASK6Q33FPXNqEGLVhRdvKe2jtV+gtt9e815nqmuFeW6BFX60E5AAbIMW2yBLyI5s
+vuTTJVzAaNtkn32HhW2m5W94NTCL/RVYJTIJebn2G+STfNtuvT7aPazT45MwtHJl
+vDexRBOs1VEu2CadT9af6JIa+ba1rLQbLY9lClsIgQKCAQEAzJr44FHQqFRggh6e
+mdYkUqnsaRlgveCcBbU7G3ZBudech/D1gMgHSNbNYVkyE3wH41zAY39zdpal3l7D
+5hapdzY/KlPODAFSjDVYyBBFTNVsQjOYOeupCQIcSTSIvI9Jv3i4+QiWKjsAuKPZ
+J6DwAxwXl5c83omSAMXQVgqBoPHLMm1Rb3m56BN6yMNrLTpIe1vLNYOCI5+mz5nW
+dqThdwH9W7W09GWTD5BloFuNrMnDJh7c1J89fKPKclGOwQCpPu2O2ER/g0zCsd5E
+/2nxGwVRtSl9JTmXmE1c6SryEipew24FH7AzsLWU0AR+h6/HLffxC5N/NwGdID84
+4vz9NwKCAQBm/2oczQQvYLcbUxIEuSl/Frg86RaN9ZLyNJg05++P9UfD6Fe3lo7o
+CplqaTyZG1GNIKIT1/u7V2Jg4SV/JxDA/rNkNfQaktTkGcpvvLZSaByGmuo4gP9i
+Wcac+xCxf3ypsGGYDoNYL227DiC0LFBVB2P73ONOnH3RyA9Jt7wWN2Lfl1JSooYP
+H71a7J/yE6GTnJIktMAALU1FAarlJxbbEdvqC8AtTFXnPqiuoDtlqc2nL73uFJih
+PyglZ9XzZpvek43gOqNQOcUcaRx38B08RPqWbvfEnQvAJVbS5jRQkdAqF477fR66
+h0xtEm3UdYCtXEXWTTMuVK5WicLNmVgBAoIBACF6xdGPpo+SD3KSU1YaJ2C149ev
+l6mLqAU2tZTMIOJLMT/gEA28Csrry7c123xbXGUBfQdw3mK978RxsIVwBZlrozK7
+XOKOA2Cu31PdztVM5woDZKqMgWUahaLVFu91Kn5th6MEIeU81mK/SkPi5jwQpv+D
+JqZSH/Eht2lAgBPK7pVRMYU0F/buPmSiVtZZfP9IJrtxgXl2irnpX+PCCn5RHvfS
+ZfClhGbVuFPVZLdjE7yU3cXG89tx4V1E4OXc+Shc642saf/q5ubp4fhip7hm17An
+0wnfRuCGgHZrTORtpWslNcezgUjLvdGAaK2cTx27s3/tCUTT96fvLS8XZF0=
+-----END RSA PRIVATE KEY-----
+EOT
+}
+
+variable "public_key" {
+  description = "The base64-encoded public certificate used by the agent"
+  default = "MIICCgKCAgEAuDZwKNah3q5LsCBmPfokvwV+cMU/Bv57DnI5KQtMDPDipAsgJhb9f0TFSoSXTA9bbG+nCQca0npf03Saes4saFxEy8hP9tpRcyYDZvYKpLP7AzXlOkvwf5QqNRFLKYT0VSX069UHhQcGOzYL0CZL2bTtlLREIQNNH1kFW5sjGgrpEsMWDuKFGou2iqgWQEna6p/O/inxcnWERLbrfU+IfstK3BCHnKUAsno0ERUh/LyQrYj+svJYESGaiHoHuP6picRrNDV5V1ntbkYTV17e/LOGSEocp/Gp9t+PZh8Rf8+ZjouCIdnJo+/lIX5XPsTigKwbbEtqBCLx3S+0qjti4Afx/IV4ZAB9KMyOjfSM+ELPU080tI7cIU ...
+}
```
We shouldn't be uploading any private/public key pairs, these should be generated or provide guidance on how to generate in the example

stephybun commented on internal/services/arckubernetes/arc_kubernetes_cluster_resource.go:
```diff
@@ -0,0 +1,232 @@
+
+	id := arckubernetes.NewConnectedClusterID(subscriptionId, d.Get("resource_group_name").(string), d.Get("name").(string))
+
+	if d.IsNewResource() {
```
Apologies for missing this in the previous reviews but this probably isn't required since the Create/Update methods have been split

stephybun commented on internal/services/arckubernetes/arc_kubernetes_cluster_resource.go:
```diff
@@ -0,0 +1,232 @@
+	"github.com/hashicorp/go-azure-helpers/resourcemanager/commonschema"
+	"github.com/hashicorp/go-azure-helpers/resourcemanager/identity"
+	"github.com/hashicorp/go-azure-helpers/resourcemanager/location"
+	tagsHelper "github.com/hashicorp/go-azure-helpers/resourcemanager/tags"
+	arckubernetes "github.com/hashicorp/go-azure-sdk/resource-manager/hybridkubernetes/2021-10-01/connectedclusters"
+	"github.com/hashicorp/terraform-provider-azurerm/helpers/tf"
+	azValidate "github.com/hashicorp/terraform-provider-azurerm/helpers/validate"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/clients"
+	"github.com/hashicorp/terraform-provider-azurerm/internal/tags"
```
We should be able to remove the old tags import 
```suggestion
	"github.com/hashicorp/go-azure-helpers/resourcemanager/tags"
	arckubernetes "github.com/hashicorp/go-azure-sdk/resource-manager/hybridkubernetes/2021-10-01/connectedclusters"
	"github.com/hashicorp/terraform-provider-azurerm/helpers/tf"
	azValidate "github.com/hashicorp/terraform-provider-azurerm/helpers/validate"
	"github.com/hashicorp/terraform-provider-azurerm/internal/clients"
```

stephybun commented on internal/services/arckubernetes/testdata/install_agent.py:
```diff
@@ -0,0 +1,249 @@
+                         args.location,
+                         args.tenantId,
+                         privateKey,
+                         # private_key_pem2,
```
```suggestion

```

stephybun commented on website/docs/r/arc_kubernetes_cluster.html.markdown:
````diff
@@ -0,0 +1,106 @@
+  }
+}
+```
+
````
It might be helpful to add a note to point users to the example, something like
```suggestion

-> **Note:** An extensive example on connecting the `azurerm_arc_kubernetes_cluster` to an external kubernetes cluster can be found in [the `./examples/arckubernetes` directory within the GitHub Repository](https://github.com/hashicorp/terraform-provider-azurerm/tree/main/examples/arckubernetes).
```

stephybun reviewed (changes requested):
Thanks for making those updates and for adding the complete example @ms-zhenhua. A few final remarks but it looks like this will be good to go once those have been addressed.

stephybun reviewed (approved):
Thanks for your effort on this @ms-zhenhua. LGTM 🥳
